JavaScript rewrite & cleanup

JackuB opened the pull request:
- coffee -> js
- removed jwt.decode (as it was unsafe and basically pointless)
- rewrote almost all tests (previously they were giving false positives)
- made passport strategy actually passport compatible
- added option to query remote server for user data (it will be kotbod in our case)
- updated few dependecies and readme

Comments are welcome

LilyJ commented on lib/environment.js:
```diff
@@ -0,0 +1,16 @@
+  }
+
+  return process.env.AUTHSERVER;
+}
```
@JackuB FWIW, you can replace all `function () {...}` with fat arrow ;)

JackuB replied:
Agreed, but you need to enable it with flag. Seems like it's not there yet :/ (and I'm not feeling like doing custom builds of node :D)

![snimek obrazovky 2015-04-03 v 2 03 42](https://cloud.githubusercontent.com/assets/1788727/6976146/b4f28e6e-d9a5-11e4-8d4e-b26dfcff857a.png)

LilyJ replied:
That's a bummer. I guess we can't avoid using node.

JackuB replied:
Seems like full (and native) support is still in far future: https://github.com/joyent/node/wiki/ES6-(a.k.a.-Harmony)-Features-Implemented-in-V8-and-Available-in-Node

LilyJ commented on lib/jwt-utils.js:
```diff
@@ -0,0 +1,60 @@
+var _ = require("underscore");
+var jsonwebtoken = require('jsonwebtoken');
+var environment = require('./environment');
```
nit: please be consistent with single vs double quotes.

JackuB replied:
Fixed across all files

LilyJ commented on lib/jwt-utils.js:
```diff
@@ -0,0 +1,60 @@
+var _ = require("underscore");
+var jsonwebtoken = require('jsonwebtoken');
```
Hmm.. should `jsonwebtoken` be `jsonWebToken`?

JackuB replied:
Renamed to `jwt` - same as auth0 use in their docs.

LilyJ commented on lib/jwt-utils.js:
```diff
@@ -0,0 +1,60 @@
+  if(typeof options === "function") {
+    done = options;
+    options = {};
+  }
```
Hmm.. I guess you are trying to check if there options is there, but this hard to read. I'd vote for checking arguments instead, like:

```
if (arguments.length === 3) {
  var done = arguments[2];
}
```

JackuB replied:
I've tried to make it shorter, but 4 lines is best I could get. Changed condition to arg count.

LilyJ commented on lib/jwt-utils.js:
```diff
@@ -0,0 +1,60 @@
+module.exports.verify = function(token, publicKey, options, done) {
+  if(typeof options === 'function') {
+    done = options;
+    options = {};
```
ditto

LilyJ commented on lib/jwt-utils.js:
```diff
@@ -0,0 +1,60 @@
+    }
+    return done(err);
+  }
+};
```
Looks like there are a lot of similarity in `sign` and `verify`, could we dry it?

JackuB replied:
Haven't really found that much similarities (except argument normalization, which can't be simplified much more IMO - and DRYing it calls for ugliness, as `undefined` part of `arguments` can't be changed).

LilyJ commented on lib/strategies.js:
```diff
@@ -0,0 +1,53 @@
+  return function(token, done) {
+    jwt.verify(token, publicKey, function(err, payload) {
+      if (err) {
+        done(err);
```
missing return

JackuB replied:
`jwt.verify` is async, so return would be pointless there.

LilyJ replied:
Oh yeah, you are right.

LilyJ commented on lib/strategies.js:
```diff
@@ -0,0 +1,53 @@
+      if (err) {
+        done(err);
+      }
+      if(environment.authserver()) {
```
What does environment.authserver() return?

JackuB replied:
Either `false` (for not found or non-https) or string with remote server address.

LilyJ commented on package.json:
```diff
@@ -19,9 +19,10 @@
-    "passport-http-bearer": ""
+    "passport-http-bearer": "",
+    "request": "^2.54.0",
+    "underscore": "^1.8.2"
```
Lo-Dash is better. I believed we use that in our main app as well.

JackuB replied:
Okay, you are right, switched.

LilyJ commented on test/lib/jwt-utils.spec.js:
```diff
@@ -0,0 +1,86 @@
+var specHelper = require('../spec-helper');
+var expect = specHelper.expect;
+var sinon = specHelper.sinon;
```
You can utilize ES6's destructuring :)

JackuB replied:
Not in nodejs yet. Request raised 5 years ago O.o https://code.google.com/p/v8/issues/detail?id=811

LilyJ replied:
wahhh...really? So we can't all the cool ES6 stuff :(

JackuB replied:
I got this running with https://babeljs.io

```javascript
exports.verbose = () => {
  return !!parseInt(process.env.VERBOSE, 10);
};
```

but `Not suitable for libraries` https://babeljs.io/docs/usage/require/

LilyJ commented on test/lib/jwt-utils.spec.js:
```diff
@@ -0,0 +1,86 @@
+var authHelper = require('../support/authentication-helper');
+var oauthKey = authHelper.oauthKey;
+var oauthPublicKey = authHelper.oauthPublicKey;
+var clientPublicKey = authHelper.clientPublicKey;
```
ditto

LilyJ commented on test/lib/jwt-utils.spec.js:
```diff
@@ -0,0 +1,86 @@
+  });
+
+  describe('#sign', function() {
+    context('with valid rsa private key', function() {
```
nit: uppercase RSA

vcelari commented on test/lib/jwt-utils.spec.js:
```diff
@@ -0,0 +1,86 @@
+          done();
+        });
+      });
+      it('JWT can be verified', function(done) {
```
This should start with a present tense verb and state what happens. See the wording in the coffeescript tests for examples. Here and elsewhere, e.g. `it('builds a valid token with the correct payload'), etc.

If you read the out put of the unit test, it should read as documentation telling you exactly what happened. 

Though this is rspec, this is solid advice about unit tests: http://blog.carbonfive.com/2010/10/21/rspec-best-practices/

LilyJ replied:
:+1:

JackuB replied:
Good link. Updated few descriptions.

LilyJ commented on test/lib/jwt-utils.spec.js:
```diff
@@ -0,0 +1,86 @@
+      context('with an invalid public RSA public key', function() {
+        it('ends with err and without payload', function(done) {
+          jwt.verify(_this.token, clientPublicKey, function(err, payload) {
+            expect(err).to.exist;
```
Do you throw specific error for this? If yes, you might as well test the expected err

JackuB replied:
Good idea, added those tests.

LilyJ commented on test/lib/jwt-utils.spec.js:
```diff
@@ -0,0 +1,86 @@
+    context('with an invalid token', function() {
+      it('ends with err and without payload', function(done) {
+        jwt.verify('abc.def.', oauthPublicKey, function(err, payload) {
+          expect(err).to.exist;
```
ditto

LilyJ commented on test/lib/strategies.spec.js:
```diff
@@ -0,0 +1,85 @@
+      _this.info = sinon.mock('info');
+      _this.verifyPayloadCallback = sinon.stub().returns(null, _this.user, _this.info);
+      _this.strategyFunction = new strategies.verifyTokenStrategy(clientPublicKey);
+    });
```
I think if you mocked, you have to restore it. Otherwise, you just leave the garbage here, right?

JackuB replied:
I've found in docs that you need to restore only for mocked servers/ajax calls. But not totally sure about it.

LilyJ replied:
Worth finding out then

JackuB replied:
Tests acts same, Sinon docs doesn't mention it that much. Seems like it should be used only for repeating assertions on calledOnce etc.

LilyJ commented on test/lib/strategies.spec.js:
```diff
@@ -0,0 +1,85 @@
+      var strategyFunction = new strategies.verifyTokenStrategy(oauthPublicKey);
+      it('ends with err and without payload', function(done) {
+        strategyFunction(_this.token, function(err, payload) {
+          expect(err).to.exist;
```
ditto

LilyJ commented on test/lib/strategies.spec.js:
```diff
@@ -0,0 +1,85 @@
+      var strategyFunction = new strategies.verifyTokenStrategy(clientPublicKey);
+      it('ends with err and without payload', function(done) {
+        strategyFunction('a.bogus.token', function(err, payload) {
+          expect(err).to.exist;
```
ditto

LilyJ commented on test/support/authentication-helper.js:
```diff
@@ -0,0 +1,49 @@
+var _ = require('lodash');
+var fs = require('fs');
```
What's fs?

JackuB replied:
[Filesystem](https://nodejs.org/api/fs.html)

LilyJ replied:
Please be mindful of your colleagues and be explicit with variable name.

smizell replied:
Sorry to jump in, just saw in my email. The `fs` library is a Node.js library and is always used as `fs` in the Node.js ecosystem as far I can tell.

LilyJ replied:
True. Poo brain.

LilyJ commented on test/support/authentication-helper.js:
```diff
@@ -0,0 +1,49 @@
+var keyPath = __dirname + '/../fixtures';
+
+var ref = ['oauth', 'client'];
+for (var i = 0; i < ref.length; i++) {
```
nit: you can use let instead of var (if node supposed that)

JackuB replied:
Throws unknown identifier right now.

LilyJ commented on test/support/authentication-helper.js:
```diff
@@ -0,0 +1,49 @@
+for (var i = 0; i < ref.length; i++) {
+  exports[ref[i] + 'Key'] = fs.readFileSync(keyPath + '/' + ref[i] + '-dev.key');
+  exports[ref[i] + 'PublicKey'] = fs.readFileSync(keyPath + '/' + ref[i] + '-dev.key.pub');
+}
```
nit: you might want to utilize lodash each/map, since you already required it. I like underscore, but vanilla is fine as well. I'll leave it up to you.

JackuB replied:
Yup, `_.each()` looks better - changed.

LilyJ commented on test/support/authentication-helper.js:
```diff
@@ -0,0 +1,49 @@
+var verifyAccessToken = function(key, token, options, done) {
+  if (typeof options === 'function') {
+    done = options;
+    options = {};
```
ditto about arguments

LilyJ commented:
@JackuB Done reviewing. Since I'm not super familiar with this, I'll assume that you know what you are doing :P I tried my best pointing out pieces of code that I think could be improve. Hope it's helpful to you. Let me know if you have any questions.

JackuB commented:
Finished addressing comments.

BTW 
1) code reviewing on Github seems weird. Can't mark comment as resolved/blocking
2) how about we use some code quality tools? I have really good experience with [CodeClimate](https://codeclimate.com) and [Codacy](https://www.codacy.com) (both do almost the same): it checks all PRs/commits for errors, typos, cyclomatic complexity etc. (it would catch few of issues Lilly found)

Ideas? I will setup CodeClimate for some of our repos to see what it can do for us - if no one is against it.

@apiaryio/enterprise

smizell commented:
@JackuB Would you care to add it to the retro discussion we have tomorrow (Tuesday)? Seems like something we could do to improve things (but will need to be explored since it's only free for open source projects it seems).

fosrias commented on test/lib/strategies.spec.js:
```diff
@@ -0,0 +1,85 @@
+describe('Strategies Module', function() {
+  describe('#registerBearerStrategy', function() {
+    var _this = this;
+    before(function() {
```
Blank line before before and after blocks.

fosrias replied:
for readablility

fosrias commented on test/lib/strategies.spec.js:
```diff
@@ -0,0 +1,85 @@
+      _this.strategyVerificationFunction.restore();
+      _this.bearerStrategy.restore();
+    });
+    it('wraps a public key and verification callback in a function', function() {
```
Blank line before each example (it) for readability

fosrias commented on test/lib/strategies.spec.js:
```diff
@@ -0,0 +1,85 @@
+      _this.verifyPayloadCallback = sinon.stub().returns(null, _this.user, _this.info);
+      _this.strategyFunction = new strategies.verifyTokenStrategy(clientPublicKey);
+    });
+    context('with a valid token', function() {
```
blank line before context if not preceded by context or describe

fosrias commented on lib/environment.js:
```diff
@@ -0,0 +1,21 @@
+exports.verbose = function() {
```
What do you think about #isVerbose to make it clear it is a predicate method?

fosrias commented on lib/environment.js:
```diff
@@ -0,0 +1,21 @@
+ * Check whether Auth server variable is set
+ * @return {mixed} False or correct server URL
+ */
+exports.authserver = function() {
```
What do you think about a more self documenting name, like #authserverUrl?

fosrias commented:
@JackuB Few minor comments and should be good to merge. Feel free to have @tu1ly do a final review so you don't block around me to review it.

fosrias commented:
@JackuB Much improved, BTW.

JackuB commented:
Submitted edits.

@tu1ly could you pls look at this?
